src/libqof/qof/qofsession.cpp

```diff
@@ -124,528 +81,364 @@ qof_backend_get_registered_access_method_list(void)
         {
-            /* More than one backend could provide this
-            access method, check file type compatibility. */
-            type_check = (gboolean (*)(const char*)) prov->check_data_type;
-            if (type_check)
+            gboolean type_check_result = (type_check) (m_book_id.c_str ());
+            bool prov_type {type_check_result != 0};
+            if (!prov_type)
             {
-                prov_type = (type_check)(session->book_id);
-                if (!prov_type)
-                {
-                    PINFO(" %s not usable", prov->provider_name);
-                    p = p->next;
-                    continue;
-                }
-            }
-            PINFO (" selected %s", prov->provider_name);
-            if (NULL == prov->backend_new)
-            {
-                p = p->next;
+                PINFO(" %s not usable", prov->provider_name);
                 continue;
             }
-            /* Use the providers creation callback */
-            session->backend = (*(prov->backend_new))();
-            session->backend->provider = prov;
-            /* Tell the book about the backend that they'll be using. */
-            qof_book_set_backend (session->book, session->backend);
-            LEAVE (" ");
-            return;
         }
-        p = p->next;
+        PINFO (" selected %s", prov->provider_name);
+        if (!prov->backend_new)
+        {
+            PINFO ("  no constructor defined.");
+            continue;
+        }
+        /* Use the providers creation callback */
+        m_backend = (*(prov->backend_new))();
+        m_backend->provider = prov;
+        /* Tell the book about the backend that he'll be using. */
+        qof_book_set_backend (m_book, m_backend);
+        LEAVE (" ");
+        return;
     }
-    msg = g_strdup_printf("failed to load '%s' using access_method", access_method);
-    qof_session_push_error (session, ERR_BACKEND_NO_HANDLER, msg);
-    g_free(msg);
+    std::string msg {"failed to get_backend using access method \"" + access_method + "\""};
+    push_error (ERR_BACKEND_NO_HANDLER, msg);
     LEAVE (" ");
 }
 
-/* ====================================================================== */
-
-static void
-qof_session_destroy_backend (QofSession *session)
+void
+QofSessionImpl::load (QofPercentageFunc percentage_func) noexcept
 {
-    g_return_if_fail (session);
+    if (!m_book_id.size ()) return;
+    ENTER ("sess=%p book_id=%s", this, m_book_id.c_str ());
 
-    if (session->backend)
-    {
-        /* clear any error message */
-        char * msg = qof_backend_get_message (session->backend);
-        g_free (msg);
+    /* At this point, we should are supposed to have a valid book
+    * id and a lock on the file. */
+    QofBook * oldbook {m_book};
+    /* XXX why are we creating a book here? I think the books
+    * need to be handled by the backend ... especially since
+    * the backend may need to load multiple books ... XXX. FIXME.
```

Comment:
I didn't put either of those comments, but I didn't know enough to remove them. Perhaps I can remove them now.